optuna/storages/_rdb/storage.py

```diff
@@ -1132,6 +1143,54 @@ def get_all_versions(self) -> List[str]:
 
         return self._version_manager.get_all_versions()
 
+    def record_heartbeat(self, trial_id: int) -> None:
+        # The `record_heartbeat` function is called in the spawned threads.
+        # We need to create the tables in each spawned thread.
+        # See　https://github.com/optuna/optuna/pull/2190#discussion_r560727140.
+        models.BaseModel.metadata.create_all(self.engine)
+
+        with _create_scoped_session(self.scoped_session, True) as session:
+            heartbeat = models.TrialHeartbeatModel.where_trial_id(trial_id, session)
+            if heartbeat is None:
+                heartbeat = models.TrialHeartbeatModel(trial_id=trial_id)
+                session.add(heartbeat)
+            else:
+                heartbeat.heartbeat = session.execute(func.now()).scalar()
+
+    def fail_stale_trials(self) -> List[int]:
+        stale_trial_ids = self._get_stale_trial_ids()
+
+        for trial_id in stale_trial_ids:
+            self.set_trial_state(trial_id, TrialState.FAIL)
+
+        return stale_trial_ids
+
+    def _get_stale_trial_ids(self) -> List[int]:
+        assert self.heartbeat_interval is not None
+        grace_period = self.grace_period or 2 * self.heartbeat_interval
```

Comment:
We should be careful with implicit casts of int to boolean. If `self.grace_period` is `0`, it will be cast to `False` which isn't necessarily intended.